### PR TITLE
[SPARK-32292][SPARK-32252][INFRA] Run the relevant tests only in GitHub Actions

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR mainly proposes to run only relevant tests just like Jenkins PR builder does. Currently, GitHub Actions always run full tests which wastes the resources.

In addition, this PR also fixes 3 more issues  very closely related together while I am here.

1. The main idea here is: It reuses the existing logic embedded in `dev/run-tests.py` which Jenkins PR builder use in order to run only the related test cases.

2. While I am here, I fixed SPARK-32292 too to run the doc tests. It was because other references were not available when it is cloned via `checkout@v2`. With `fetch-depth: 0`, the history is available.

3. In addition, it fixes the `dev/run-tests.py` to match with `python/run-tests.py` in terms of its options. Environment variables such as `TEST_ONLY_XXX` were moved as proper options. For example,

    ```bash
    dev/run-tests.py --modules sql,core
    ```

    which is consistent with `python/run-tests.py`, for example,

    ```bash
    python/run-tests.py --modules pyspark-core,pyspark-ml
    ```

4. Lastly, also fixed the formatting issue in module specification in the matrix:

    ```diff
    -            network_common, network_shuffle, repl, launcher
    +            network-common, network-shuffle, repl, launcher,
    ```

    which incorrectly runs build/test the modules.

### Why are the changes needed?

By running only related tests, we can hugely save the resources and avoid unrelated flaky tests, etc.
Also, now it runs the doctest of `dev/run-tests.py` properly, the usages are similar between `dev/run-tests.py` and `python/run-tests.py`, and run `network-common`, `network-shuffle`, `launcher` and `examples` modules too.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

Manually tested in my own forked Spark:

https://github.com/HyukjinKwon/spark/pull/7
https://github.com/HyukjinKwon/spark/pull/8
https://github.com/HyukjinKwon/spark/pull/9
https://github.com/HyukjinKwon/spark/pull/10
https://github.com/HyukjinKwon/spark/pull/11
https://github.com/HyukjinKwon/spark/pull/12

### CODE REVIEW

File: dev/run-tests.py [hunk 1]
```diff
@@ -589,43 +627,74 @@ def main():
...
-        str_test_modules = [m.strip() for m in os.environ.get("TEST_ONLY_MODULES").split(",")]
+        str_test_modules = [m.strip() for m in opts.modules.split(",")]
         test_modules = [m for m in modules.all_modules if m.name in str_test_modules]
-        # Directly uses test_modules as changed modules to apply tags and environments
-        # as if all specified test modules are changed.
+
+        # If we're running the tests in Github Actions, attempt to detect and test
+        # only the affected modules.
+        if test_env == "github_actions":
+            if os.environ["GITHUB_BASE_REF"] != "":
```
Comment (HyukjinKwon) [inline dev/run-tests.py:656]: `GITHUB_BASE_REF` is only set in the PR builder; otherwise, it's an empty string.

File: dev/run-tests.py [hunk 2]
```diff
@@ -589,43 +627,74 @@ def main():
...
+                # Pull requests
+                changed_files = identify_changed_files_from_git_commits(
+                    os.environ["GITHUB_SHA"], target_branch=os.environ["GITHUB_BASE_REF"])
```
Comment (HyukjinKwon) [inline dev/run-tests.py:659]: `GITHUB_SHA` is the commit in PR. `GITHUB_BASE_REF` is the target repo's branch.

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -29,7 +29,7 @@ jobs:
         modules:
           - |-
             core, unsafe, kvstore, avro,
-            network_common, network_shuffle, repl, launcher
+            network-common, network-shuffle, repl, launcher,
```
Comment (dongjoon-hyun) [inline .github/workflows/master.yml:32]: Oops.

File: dev/run-tests.py (hunk 1 above)
Comment (dongjoon-hyun) [reply dev/run-tests.py:656]: Do you mean Jenkins PR Builder?

File: dev/run-tests.py [hunk 4]
```diff
@@ -589,43 +627,74 @@ def main():
...
     if should_only_test_modules:
-        str_test_modules = [m.strip() for m in os.environ.get("TEST_ONLY_MODULES").split(",")]
+        str_test_modules = [m.strip() for m in opts.modules.split(",")]
         test_modules = [m for m in modules.all_modules if m.name in str_test_modules]
-        # Directly uses test_modules as changed modules to apply tags and environments
-        # as if all specified test modules are changed.
+
+        # If we're running the tests in Github Actions, attempt to detect and test
+        # only the affected modules.
+        if test_env == "github_actions":
```
Comment (dongjoon-hyun) [inline dev/run-tests.py:655]: Thanks. This looks safe.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thanks, @HyukjinKwon .
Merged to master.

**Comment by HyukjinKwon** [comment]
Thank you @dongjoon-hyun!

File: dev/run-tests.py (hunk 2 above)
Comment (viirya) [reply dev/run-tests.py:659]: For many commits in a PR, each time `changed_files` is different? If a previous commit changed two modules but later one changes only one, will this only run test for the later module?

Comment (HyukjinKwon) [reply dev/run-tests.py:659]: From my testing, it combines all commits and `changed_files` is based on all commits always.

Seems like `GITHUB_SHA` became a merge commit internally in Github Actions, and `GITHUB_BASE_REF` becomes the branch which PR opens against.

For example, if a PR has commits 1, 2 and 3, GitHub Actions looks creates a merge commit 4 for commits 1, 2 and 3. And, it calculates the diff from the branch, for example, `master` in the original repo, Apache Spark.

```bash
git diff 4 master
```

Comment (HyukjinKwon) [reply dev/run-tests.py:659]: This is an example of the merge commit: https://github.com/HyukjinKwon/spark/commit/8f36ec455e19dbfb10195d872a9ccaeb2de8ceca at https://github.com/HyukjinKwon/spark/pull/7

Comment (viirya) [reply dev/run-tests.py:659]: Okay. Thanks for clarifying. Looks good.

**Pull request closed without merging**